myia/abstract/infer.py

```diff
@@ -296,9 +296,23 @@ def get_inferrer_for(self, m: MacroFunction):
         argrefs = [self.ref(node, ctx) for node in n_args]
 
         if isinstance(fn, AbstractType):
+            # Replace abstract type instantiation with
+            # either a cast for abstract scalars, or
+            # a make_record for all other cases.
+            val = fn.xvalue()
+            cls = type_to_abstract(val)
             g = ref.node.graph
-            newfn = g.apply(P.partial, P.make_record, fn.xvalue())
-            newcall = g.apply(newfn, *n_args)
+            if isinstance(cls, AbstractScalar):
+                typ = cls.xtype()
+                if issubclass(typ, xtype.Number):
+                    newcall = g.apply(P.scalar_cast, *n_args, cls)
+                elif typ is xtype.Bool:
+                    newcall = g.apply(P.scalar_ne, *n_args, 0)
```

Comment:
Sorry, I wasn't totally clear. This should map to the `bool` operator,  which normally calls `obj.__bool__` and will produce `x != 0` on int/float.

```suggestion
                    newcall = g.apply(operations.bool, *n_args)
```

Reply:
OK !

PS: I am also trying to write the macro `call_object`, but I got other issues. I call the macro in infer code using `g.apply(operations.call_object, fn, *n_args)`, but then it wrap the abstractype into another abstract type again (ie., macro receives `AbstractType(AbstractType(AbstractScalar))` instead of just `AbstractType(AbstractScalar)`). I'm trying to figure out why this happens.

Reply:
I assume that's because it's wrapping it with `to_abstract`. Try using `n_fn` instead of `fn`.

Reply:
Thanks !